[EE-IR] Use reflection to avoid the need for accessors when compiling fragments

kandersen opened the pull request:
- Refactor JvmIrCodegenFactory to allow for caller's choice of
  lowerings.

- Refactor of SyntheticAccessorLowering to expose the accessibility
  check used to determine the need for accessors.

- Replacement of field, method and constructor uses with corresponding
  java.lang.reflection API via additional lowering running _before_
  the existing IR pipeline.

- Emulating super calls in the context of class methods via new
  compiler intrinsic.

- psi2ir: Add support for `_field` suffix on properties in the
  evaluator for accessing underlying fields when applicable.

- Add `JvmGeneratorExtensions` point for determining whether a
  property accessor has a user-supplied body in order to predict how
  they are ultimately compiled for the JVM.

kandersen commented:
Still draft mode, but Github does not check for mergeability of drafts :sweat_smile:

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,503 @@
+            setField.symbol,
+        )
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
```
@udalov This is just about ready for review, barring this little remaining obstacle :)

The information of whether the accessor is a default accessor is lost in translation between front- and back-end, where PropertyAccessorDescriptors are mapped to IrSimpleFunctions, which do not contain the information.

A possible place to put the information may be the `correspondingPropertySymbol?.owner` of `accessor`, as a property on `IrProperty`.

What do you think?

udalov replied:
Answered in internal Slack channel

kandersen commented:
Ready for review!

udalov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/JvmDebuggerInvokeSpecial.kt:
```diff
@@ -0,0 +1,61 @@
+        return MaterialValue(codegen, Type.getReturnType(descriptor), expression.type)
+    }
+
+    private fun IrExpression.getBooleanConst() =
```
Let's remove `getBooleanConst` & `getStringConst`, the utilities with the same name already exist and can be used instead

kandersen replied:
Oh, didn't notice! I took 'em from JvmInvokeDynamic.kt, but that file is in scope here.

Should we bump them to a common utility class, or are you fine with using them from `JvmInvokeDynamic`?

udalov replied:
I have a separate unrelated change where I move declarations around, including these functions. So let's not move them here

kandersen replied:
Good, I will simply remove the utils from here, then :)

udalov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmSymbols.kt:
```diff
@@ -60,6 +60,7 @@ class JvmSymbols(
     private val kotlinReflectPackage: IrPackageFragment = createPackage(FqName("kotlin.reflect"))
     private val javaLangPackage: IrPackageFragment = createPackage(FqName("java.lang"))
     private val javaLangInvokePackage: IrPackageFragment = createPackage(FqName("java.lang.invoke"))
+    private val javaLangReflectPackage: IrPackageFragment = createPackage(FqName("java.lang.reflect"))
```
I wonder if it makes sense to extract `java.lang.reflect` API to a separate class ~`JvmReflectSymbols` and create it on-demand in `ReflectiveAccessLowering`. It'd be useful both for readability (`JvmSymbols` is very big and slow to analyze already) and performance (creating numerous IR declarations which won't be used in 99% of scenarios seems unnecessary).

But it's not a blocker for this review. We can think about it separately, together with maybe extracting some other groups of symbols out of this class with the same purpose.

kandersen replied:
Makes sense to me. I will push a separate PR with a start on this, then we can iterate on that there.

OK?

udalov replied:
Yeah, thanks, I think it can be postponed

udalov commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/CallGenerator.kt:
```diff
@@ -202,20 +202,35 @@ class CallGenerator(statementGenerator: StatementGenerator) : StatementGenerator
+                        dispatchReceiverValue.load(),
+                        IrStatementOrigin.GET_PROPERTY,
+                        superQualifierSymbol
+                    ).also { context.callToSubstitutedDescriptorMap[it] = descriptor }
```
Probably doesn't matter right now, but this `also { ... }` call looks like it should apply to the IrGetFieldImpl below too, so maybe should be extracted out of the `if` statement

kandersen replied:
Done.

udalov reviewed (changes requested):
The main implementation code looks good, I've left a few comments related mostly to not very significant parts

kandersen commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/StubGeneratorExtensions.kt:
```diff
@@ -35,6 +34,15 @@ open class StubGeneratorExtensions {
+    // bodies, information that is _not_ present in the IR structure, but _is_
+    // available in the corresponding PSI. See `CodeFragmentCompiler` in the
+    // plug-in for the implementation.
+    open fun accessorWithExplicitImplementation(accessor: IrSimpleFunction): Boolean = false
```
Done.

kandersen commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/expressions/IrStatementOrigin.kt:
```diff
@@ -121,3 +121,13 @@ fun IrStatementOrigin.isAssignmentOperatorWithResult() =
+        IrStatementOrigin.MULTEQ,
+        IrStatementOrigin.PERCEQ -> true
+        else -> isAssignmentOperatorWithResult()
+    }
```
Done.

kandersen commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/expressions/IrStatementOrigin.kt:
```diff
@@ -121,3 +121,13 @@ fun IrStatementOrigin.isAssignmentOperatorWithResult() =
+fun IrStatementOrigin.isAssignmentOperator(): Boolean =
+    when (this) {
+        IrStatementOrigin.EQ,
+        IrStatementOrigin.PLUS,
```
Good catch, thanks.

kandersen commented on compiler/ir/backend.jvm/entrypoint/src/org/jetbrains/kotlin/backend/jvm/JvmIrCodegenFactory.kt:
```diff
@@ -48,7 +48,15 @@ open class JvmIrCodegenFactory(
     private val externalSymbolTable: SymbolTable? = null,
     private val jvmGeneratorExtensions: JvmGeneratorExtensionsImpl = JvmGeneratorExtensionsImpl(configuration),
-    private val evaluatorFragmentInfoForPsi2Ir: EvaluatorFragmentInfo? = null
+    private val invokeCompilerPhases: (PhaseConfig, JvmBackendContext, IrModuleFragment) -> Unit =
```
Good idea, and that is exactly coming up in my next set of changes!

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.load.java.JvmAbi
+
+val reflectiveAccessLowering = makeIrModulePhase(
```
Done!

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+                JvmAbi.getterName(propertyName = property.name.asString()),
+                getter.extensionReceiverParameter?.let { listOf(it.type) } ?: listOf(),
+                call.dispatchReceiver!!,
+                call.extensionReceiver?.let { listOf(it) } ?: listOf(),
```
Done!

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+)
+
+// This lowering replaces member accesses that are illegal according to JVM
+// accessibility rules with corresponding calls to the java.lang.reflection
```
Good catch!

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+            return generateReflectiveMethodInvocation(
+                getter.parentAsClass.defaultType,
+                JvmAbi.getterName(propertyName = property.name.asString()),
+                getter.extensionReceiverParameter?.let { listOf(it.type) } ?: listOf(),
```
Done. Thanks, these little patterns are hard to discover, but I'm building out my vocabulary :)

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+        )
+    }
+
+    fun generateThrowIllegalAccessException(setField: IrSetField): IrExpression {
```
Done.

kandersen commented on compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:
```diff
@@ -0,0 +1,520 @@
+    private fun fieldLocationAndReceiver(call: IrCall): Pair<IrType, IrExpression?> {
+        callsOnCompanionObjects[call]?.let {
+            val parentAsClass = it.owner.parentAsClass
+            if (!parentAsClass.isInterface) {
```
Yeah, need a bit of investigation on this. Test coverage on annotations is not adequate.

kandersen replied:
Hmm, seems there's no way to force inaccessible members onto annotation classes - all members beyond public constructor arguments are rejected by the frontend.

The closest I can get is a private property on a _public_ companion object, but there the illegal access is the property on the companion, not the getting of the companion object.

So, sure, conceptually `isJvmInterface` is probably the right thing, but doesn't seem to make a difference in practice.

kandersen replied:
Added a test on the plug-in side:

https://github.com/JetBrains/intellij-community/pull/1802/files#diff-7bfba535546d72fca4ed1f138048d38899bda56a9437c93340a203e8eac79f81

kandersen commented:
Opened #4641 with extraction of `java.lang.reflect` symbols from `JvmSymbols.kt`.

udalov commented:
Some tests are failing, could you please investigate?
```
org.jetbrains.kotlin.test.runners.ir.IrTextTestGenerated$Types$NullChecks
    testImplicitNotNullOnPlatformType
    testPlatformTypeReceiver
    testTypeParameterWithMixedNullableAndNotNullableBounds
    testTypeParameterWithMultipleNotNullableBounds
    testTypeParameterWithMultipleNullableBounds
org.jetbrains.kotlin.test.runners.ir.IrTextTestGenerated$Expressions
    testCoercionToUnit
    testEquals
    testJvmStaticFieldReference
```

kandersen commented:
Yep, on it.

kandersen commented:
> Some tests are failing, could you please investigate?
> 
> ```
> org.jetbrains.kotlin.test.runners.ir.IrTextTestGenerated$Types$NullChecks
>     testImplicitNotNullOnPlatformType
>     testPlatformTypeReceiver
>     testTypeParameterWithMixedNullableAndNotNullableBounds
>     testTypeParameterWithMultipleNotNullableBounds
>     testTypeParameterWithMultipleNullableBounds
> org.jetbrains.kotlin.test.runners.ir.IrTextTestGenerated$Expressions
>     testCoercionToUnit
>     testEquals
>     testJvmStaticFieldReference
> ```

Fixed! I was dropping some super qualifiers by accident. Removed the branching, this seems to pass all tests I can find on both compiler and plugin side.

udalov commented:
Looks good now, thank you!
